src/main/kotlin/no/nav/medlemskap/common/HttpClientMetrics.kt

```diff
@@ -0,0 +1,102 @@
+import io.ktor.client.HttpClientConfig
+import io.ktor.client.call.HttpClientCall
+import io.ktor.client.features.HttpClientFeature
+import io.ktor.client.request.HttpRequestBuilder
+import io.ktor.client.request.HttpSendPipeline
+import io.ktor.client.statement.HttpResponse
+import io.ktor.client.statement.HttpResponsePipeline
+import io.ktor.util.AttributeKey
+import io.prometheus.client.Histogram
+
+/**
+ * [HttpClient] Prometheus metrics feature.
+ */
+class HttpClientMetrics {
+
+    /**
+     * [HttpClientMetrics] feature configuration
+     */
+    class Config {}
+
+    private suspend fun logRequest(request: HttpRequestBuilder) {
+        val service = request.url.encodedPath
+        val operation = request.method.value
+        val timer = restClientTimer.labels(service, operation).startTimer()
+        request.attributes.put(timerKey, timer)
+    }
+
+    private suspend fun logResponse(response: HttpResponse) {
+        val status = response.status.value.toString()
+        val operation = response.call.request.method.value
+        val service = response.call.request.url.encodedPath
+        restClientCounter.labels(service, operation, status).inc()
+
+        response.call.attributes.getOrNull<Histogram.Timer>(timerKey)?.observeDuration()
+    }
+
+    private fun logRequestException(context: HttpRequestBuilder, cause: Throwable) {
+        val service = context.url.encodedPath
+        val operation = context.method.value
+        restClientCounter.labels(service, operation, Companion.REQUEST_EXCEPTION_STATUS).inc()
+    }
+
+    private fun logResponseException(context: HttpClientCall, cause: Throwable) {
+        val service = context.request.url.encodedPath
+        val operation = context.request.method.value
+        restClientCounter.labels(service, operation, Companion.RESPONSE_EXCEPTION_STATUS).inc()
+    }
+
+    companion object : HttpClientFeature<Config, HttpClientMetrics> {
+        override val key: AttributeKey<HttpClientMetrics> = AttributeKey("HttpClientMetrics")
+
+        override fun prepare(block: Config.() -> Unit): HttpClientMetrics {
+            return HttpClientMetrics()
+        }
+
+        override fun install(feature: HttpClientMetrics, scope: HttpClient) {
+            scope.sendPipeline.intercept(HttpSendPipeline.Monitoring) {
+                try {
+                    feature.logRequest(context)
+                } catch (_: Throwable) {
+
+                }
+
+                try {
+                    proceedWith(subject)
+                } catch (cause: Throwable) {
+                    feature.logRequestException(context, cause)
+                    throw cause
+                }
+            }
+
+            scope.responsePipeline.intercept(HttpResponsePipeline.Receive) {
+                try {
+                    proceedWith(subject)
+                    try {
+                        feature.logResponse(context.response)
+                    } catch (_: Throwable) {
+
```

Comment:
Er tom catch-block her det som er vanlig? er bare sjelden jeg ser det. ikke det at jeg vet hva som burde stå her altså :)

Men må denne være i try-catch i så fall? Skjønner den er her for å overleve eventuelle feil, men hva slags feil kan kastes?

Reply:
Jeg la inn litt info-logging der, kanskje det er bedre kotyme?